store/stat_store.go

```diff
@@ -0,0 +1,214 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"container/list"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+func NewStatStore(e float64, r time.Duration, w time.Duration, sp []float64) *statStore {
+	return &statStore{
+		buffer:               list.New(),
+		epsilon:              e,
+		resolution:           r,
+		windowDuration:       w,
+		supportedPercentiles: sp,
+	}
+}
+
+// statStore is an in-memory rolling timeseries buffer that allows the extraction of -
+// segments of the stored timeseries and derived stats.
+// The tradeoff between space and precision can be configured through epsilon and resolution.
+// statStore only retains information about the latest TimePoints that are within its
+// specified duration.
+// statStore is a TimeStore-like object that does not implement the TimeStore interface.
+type statStore struct {
+	// start is the start of the time window.
+	start time.Time
+
+	// buffer is a list of tpEntry that is sequenced in a time-descending order, meaning that
+	// Front points to the latest tpEntry and Back to the oldest one.
+	buffer *list.List
```

Comment:
this library assumes that the window size is a constant number of elements. In this case, we want an arbitrary number of tpEntries whose count fields sum up to the window size. For example, for a 1-minute resolution, 1-hour duration statStore, we might have just one tpEntry with a count of 60, if the value of the metric has not changed more than epsilon, on average.

Due to that, our corresponding rewind operation would subtract values from the count of the earliest tpEntry, and it would be removed only if the count is zero.